File: csharp/825.friends-of-appropriate-ages.cs
/*
 * @lc app=leetcode id=825 lang=csharp
 *
 * [825] Friends Of Appropriate Ages
 */

using System;

// @lc code=start
public class Solution {
    public int NumFriendRequests(int[] ages) {
        int n = ages.Length, ret = 0;
        Array.Sort(ages);
        for (int i = 0; i < n; ++i) {
            int age = ages[i];
            double threshold = 0.5 * (double)age + 7;
            for (int j = i - 1; j >= 0; --j) {
                int k = ages[j];
                if ((double)k <= threshold) {
                    break;
                }
                if (k == age) {
                    ++ret;
                }
                ++ret;
            }
        }
        return ret;
    }
}
// @lc code=end

File: java/825.friends-of-appropriate-ages.java
/*
 * @lc app=leetcode id=825 lang=java
 *
 * [825] Friends Of Appropriate Ages
 */

// @lc code=start
class Solution {
    public int numFriendRequests(int[] ages) {
        int[] count = new int[121];
        int ret = 0;
        for (int age : ages) {
            ++count[age];
        }
        for (int i = 1; i <= 120; ++i) {
            int t = count[i];
            count[i] += count[i - 1];
            if (i >= 15 && t >= 0) {
                int p = count[i] - count[i / 2 + 7];
                ret += p * t - t;
            }
        }
        return ret;
    }
}
// @lc code=end

